spawner/pkg/spawner/spawner.go

```diff
@@ -0,0 +1,198 @@
+	trueVal := true
+	freeMRU := uint64(4 * gb)
+	freeSRU := uint64(20 * gb)
+
+	filter := types.NodeFilter{
+		Status:  []string{"up"},
+		Healthy: &trueVal,
+		FreeMRU: &freeMRU,
+		FreeSRU: &freeSRU,
+		FarmIDs: []uint64{farm},
+	}
+	nodes, err := deployer.FilterNodes(ctx, tfPluginClient, filter, []uint64{freeSRU}, nil, nil)
+	if err != nil {
+		log.Fatal().Err(err).Send()
+	}
+
+	return nodes, nil
+}
+
+// spawn creates and deploys VMs on the specified nodes according to the provided configuration
+func spawn(ctx context.Context, tfPluginClient deployer.TFPluginClient, cfg Config, nodes []types.Node, vmCount int) error {
+	var networks []*workloads.ZNet
+	var vms []*workloads.Deployment
+
+	for i := 0; i < vmCount; i++ {
+		node := nodes[i]
+
+		network := workloads.ZNet{
+			Name:  fmt.Sprintf("network_%d", node.NodeID),
+			Nodes: []uint32{uint32(node.NodeID)},
+			IPRange: gridtypes.NewIPNet(net.IPNet{
+				IP:   net.IPv4(10, 20, 0, 0),
+				Mask: net.CIDRMask(16, 32),
+			}),
+			AddWGAccess:  false,
+			SolutionType: fmt.Sprintf("vm/%d", node.FarmID),
+		}
+		vm := workloads.VM{
+			Name:        fmt.Sprintf("vm_%d", node.NodeID),
+			Flist:       "https://hub.grid.tf/amryassir.3bot/benchmark.flist",
+			CPU:         2,
+			Planetary:   true,
+			Memory:      1024,
+			Entrypoint:  "/sbin/zinit init",
+			NetworkName: network.Name,
+			EnvVars: map[string]string{
+				"INFLUX_URL":    cfg.Influx.URL,
+				"INFLUX_ORG":    cfg.Influx.Org,
+				"INFLUX_TOKEN":  cfg.Influx.Token,
+				"INFLUX_BUCKET": cfg.Influx.Bucket,
+				"NODE_ID":       fmt.Sprintf("%d", node.NodeID),
+				"FARM_ID":       fmt.Sprintf("%d", node.FarmID),
+				"SSH_KEY":       cfg.SSHKey,
+			},
+		}
+		dl := workloads.NewDeployment(
+			fmt.Sprintf("vm_%d", node.NodeID),
+			uint32(node.NodeID),
+			fmt.Sprintf("vm/%d", node.FarmID),
+			nil,
+			network.Name,
+			nil,
+			nil,
+			[]workloads.VM{vm},
+			nil,
+		)
+
+		networks = append(networks, &network)
+		vms = append(vms, &dl)
+	}
+	err := tfPluginClient.NetworkDeployer.BatchDeploy(ctx, networks)
+	if err != nil {
+		err = handleFailure(ctx, err, cfg, tfPluginClient, networks, vms)
+		if err != nil {
+			return err
+		}
+	}
+	err = tfPluginClient.DeploymentDeployer.BatchDeploy(ctx, vms)
+	if err != nil {
+		err = handleFailure(ctx, err, cfg, tfPluginClient, networks, vms)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// calculateVMCount calculates the number of VMs to deploy based on the deployment strategy
+func calculateVMCount(nodes []types.Node, strategy float64) int {
+	totalNodes := len(nodes)
+
+	return int(float64(totalNodes) * strategy)
+}
+
+// handleFailure handles deployment failures according to the specified failure strategy
+func handleFailure(ctx context.Context, err error, cfg Config, tfPluginClient deployer.TFPluginClient, networks []*workloads.ZNet, vms []*workloads.Deployment) error {
```

Comment:
the way failures are handled needs some improvements as there are some problems with current implementation:
- the logic for deployments is scattered around the code so multiple places try deploying which make it not very clear
- it doesn't take into account which networks were deployed to try the remaining ones (same for vms)
- it only destroys vms but not networks
- destroy failing should destroy networks with failing vms and return

A better approach is to take into consideration what failed and the remaining workloads to be deployed.